tari-project/tari-dan
# fix(confidential)!: fix confidential transfer

Description
---
- wallet SDK: fix KDFs for encrypted value and mask to exactly match base layer
- wallet SDK: key manager does not implicitly set the last used key as active
- wallet SDK: check that shared mask is correct before masking output as Unspent
- wallet SDK: dont finalize outputs for dry-run
- walletd: fixes deadlock in account monitor

Motivation and Context
---
Relies on https://github.com/tari-project/tari/pull/5245

This PR corrects a number of bugs, makes minor logging and cli improvements and gets confidential transfers working.

How Has This Been Tested?
---
Manually, as per steps below

What process can a PR reviewer use to test or verify this change?
---
1. create 2 accounts named primary and alt

```shell
cargo run --bin tari_dan_wallet_cli --  accounts create --name primary
cargo run --bin tari_dan_wallet_cli --  accounts create --name alt
```

2. Use POSTman to burn using `primary`'s public key
3. Mine 4 blocks and claim the burn using the proofs returned in POSTman 
```shell
cargo run --bin tari_dan_wallet_cli --  accounts claim-burn -n primary
```

4. Transfer from `primary` to `alt`
```shell
cargo run --bin tari_dan_wallet_cli --  transactions confidential-transfer -n1 primary 100 $ALT_ACCOUNT $ALT_ACCOUNT_PK
```

5. Check balance

```
$ cargo run --bin tari_dan_wallet_cli --  accounts get-balances --account-name primary
Account component_b0fb082aa0b63b6d3b26f374a5e3f442257899f99fcec96002a7281e77716c04 balances:

VaultId                                                                | Resource                                                                               | Balance                                 
---------------------------------------------------------------------- | -------------------------------------------------------------------------------------- | --------------------------------------- 
vault_ac223920e139a33579e7497e5f2e1055e3100fb0e0600af686ba8f7ec768e9ff | resource_0101010101010101010101010101010101010101010101010101010101010101 Confidential | 0 revealed + 10000 blinded = 10000 tXTR 

1 row(s)

```

6. Check in `outputs` database table that appropriate outputs are marked as `Spent`/`Unspent` and their unblinded value is included

Breaking Changes
---

- [x] Requires data directory to be deleted (walletd)
